### ISSUE TITLE
$PATH is incorrect when running fish in a nix-shell

### ISSUE THREAD

**Issue by yacinehmito** [NixOS/nixpkgs#18946]
## Issue description

When starting a fish shell after dropping in a nix-shell, the `PATH` environment variable is improperly set. This makes fish unusable for my use case with nix-shell.
### Steps to reproduce

Say I have the following `default.nix` file in directory :

``` nix
with import <nixpkgs> {};
stdenv.mkDerivation {
  name = "whatever";
  buildInputs = [
    whatever-dependency
  ];
}
```

In the same directory, if I do `nix-shell`, I drop into a bash shell where the bin path of `whatever-dependency` is added to the `PATH` environment variable (if applicable).
However, if I start a fish shell from this bash instance, the `PATH` variable reverts to its default value.
The problem is the same when doing directly `nix-shell --run fish` or `nix-shell --command fish`. This doesn't cause any issue with zsh.

In the fish instance, other environment variables are properly updated. The `IN_NIX_SHELL` variable is indeed set to 1 and if you add some `shellPrompt` code in the derivation, you do inherit the proper environment. The `PATH` variable seems to be the only one affected.

I suspect that somewhere in `fish-foreign-env` we recreate the `PATH` from scratch instead of using the value of the parent shell. I might be mistaken ; I just tried fish on NixOS today and don't know much.
## Technical details
- System: 17.03.git.2b0eace (Gorilla)
- Nix version: nix-env (Nix) 1.11.4
- Nixpkgs version: "17.03.git.2b0eace"

**Comment by rasendubi** [comment]
The issue doesn't seem to be related to nix-shell, as it's reproducible outside of it:
```
rasen@Larry ~ % export PATH=/my/dir/:$PATH
rasen@Larry ~ % echo $PATH
/my/dir/:/home/rasen/.local/bin:/home/rasen/bin:/usr/local/bin:/usr/local/sbin:/home/rasen/bin:/var/setuid-wrappers:/home/rasen/.nix-profile/bin:/home/rasen/.nix-profile/sbin:/home/rasen/.nix-profile/lib/kde4/libexec:/nix/var/nix/profiles/default/bin:/nix/var/nix/profiles/default/sbin:/nix/var/nix/profiles/default/lib/kde4/libexec:/run/current-system/sw/bin:/run/current-system/sw/sbin:/run/current-system/sw/lib/kde4/libexec
rasen@Larry ~ % fish
rasen@Larry ~> echo $PATH
/home/rasen/bin /var/setuid-wrappers /home/rasen/.nix-profile/bin /home/rasen/.nix-profile/sbin /home/rasen/.nix-profile/lib/kde4/libexec /nix/var/nix/profiles/default/bin /nix/var/nix/profiles/default/sbin /nix/var/nix/profiles/default/lib/kde4/libexec /run/current-system/sw/bin /run/current-system/sw/sbin /run/current-system/sw/lib/kde4/libexec
```

The issue seems to be NixOS-specific, as nixpkgs-fish works fine outside of NixOS (Gentoo).

**Comment by therealpxc** [comment]
I've noticed a very strange problem, potentially related, in NixOS 17.03. When I use fish as a login shell at a virtual console, I get a ton of errors and login hangs. I do not get this behavior when I open a terminal in the GUI. I also do not get this behavior if I disable my oh-my-fish profile— even if I then go on to source my OMF config manually after logging in at the virtual console.

The particular errors I see from OMF, though, are about binaries (system dependencies of my OMF plugins) being missing, which sounds very much like a path problem.

I still don't really know going on, but the fish-foreign-env plugin we are using to run NixOS's POSIX sh scripts to set the appropriate fish environment variables is quite out of date.

The [plugin](https://github.com/oh-my-fish/plugin-foreign-env) has had several bug-fixing commits since we last updated it (January 2015), including

- [allowing '=' signs in variables](https://github.com/oh-my-fish/plugin-foreign-env/commit/25141a49be5631b07bc1321e0ea33ac341d623b8)
 - [allowing parentheses in variable values](https://github.com/oh-my-fish/plugin-foreign-env/commit/6a236cc8d85da5bb7c4316fe05e5ee78663853fe)
 - [handling empty inputs correctly](https://github.com/oh-my-fish/plugin-foreign-env/commit/03df46840f30ccd9ec9bf9070ed68238c12bfc43)
 - [correctly handling multiline vars](https://github.com/oh-my-fish/plugin-foreign-env/commit/754edaf059eea6717514a7af31e1dfc49a0836f6)
 - [handling '%s' correctly in vars](https://github.com/oh-my-fish/plugin-foreign-env/issues/3)
 - [modifying variables only for the currently running shell, not globally](https://github.com/oh-my-fish/plugin-foreign-env/issues/9)

Any of those bugs could cause some weird, bad behavior.

I'm going to update our fish-foreign-env package right now and see where that takes us.

**Comment by therealpxc** [comment]
That didn't seem to help. I am now able to log in (looks like the omf plugin 'local-config' was causing the hard hang. Lots of things are getting errors. I checked my path and it's set to /bin and /usr/bin.

I think maybe the user config is being read before the PATH is set.

**Comment by therealpxc** [comment]
@rasendubi , are you using fish as a login shell (or do you at least have `programs.fish.enable = true;` set?

The foreign-env plugin for fish by design simply replaces the values of the environment variables in fish with whatever they were set to in the bash script, if they've changed at all.

What's happening on NixOS is that every fish invocation sources `${config.system.build.setEnvironment}` through `fenv`. The sourced file totally clobbers the PATH (as it should), and foreign-env is faithfully reproducing that clobbering (as it should). Here is the relevant line from that file on my system:

    export PATH="$HOME/.nix-profile/bin:$HOME/.nix-profile/sbin:$HOME/.nix-profile/lib/kde4/libexec:/nix/var/nix/profiles/default/bin:/nix/var/nix/profiles/default/sbin:/nix/var/nix/profiles/default/lib/kde4/libexec:/run/current-system/sw/bin:/run/current-system/sw/sbin:/run/current-system/sw/lib/kde4/libexec"

Notice that there are no instances of ‘$PATH’ on the right-hand side there.

I took a look at the module for zsh, and it looks like what we want is likely just to prevent NixOS' /etc/fish/config.fish script from being sourced multiple times by the same shell or its descendants.

**Comment by therealpxc** [comment]
I [did a thing](https://github.com/NixOS/nixpkgs/pull/24314), you guys! Let me know what you think.

**Comment by therealpxc** [comment]
It sure does, my friend!

    ~/Playground  cat ./fishie-shell.nix                                                 3698ms  Sat 25 Mar 2017 08:54:34 AM MST
    with import <nixpkgs> {};
    stdenv.mkDerivation {
      name = "whatever";
      buildInputs = [
        pkgs.psmisc # for killall
      ];
    }
    ~/Playground  which killall                                                                   Sat 25 Mar 2017 08:54:37 AM MST
    which: no killall in (/home/pxc/bin:/run/wrappers/bin:/run/wrappers/bin:/home/pxc/.nix-profile/bin:/home/pxc/.nix-profile/sbin:
    /home/pxc/.nix-profile/lib/kde4/libexec:/nix/var/nix/profiles/default/bin:/nix/var/nix/profiles/default/sbin:/nix/var/nix/profi
    les/default/lib/kde4/libexec:/run/current-system/sw/bin:/run/current-system/sw/sbin:/run/current-system/sw/lib/kde4/libexec:/ni
    x/store/sq1q54snjvwk5ka7iwdv04nn42h84r55-dbus-1.10.16/bin:/nix/store/zks6v9wdc7jp8wg0yd74b48sdsavgcgw-qttools-5.7.1/bin:/nix/st
    ore/g4kkh32q8rj390fyz1fwqqlpy2vbz3qj-socat-1.7.3.2/bin:/nix/store/avmxym1w34sc17nrilsmgrk469l3ml0z-gnugrep-3.0/bin:/nix/store/0
    xwrn1p8fp8h3cynszpgbmhmydbzhns5-gnused-4.4/bin:/nix/store/zlqp9kb8399hvihidlnzqsad3q1ivbxz-kconfig-5.31.0/bin:/nix/store/zlqp9k
    b8399hvihidlnzqsad3q1ivbxz-kconfig-5.31.0/lib/libexec:/nix/store/zlqp9kb8399hvihidlnzqsad3q1ivbxz-kconfig-5.31.0/lib/libexec/kf
    5:/nix/store/40b3czpgrs222fadr11y22phxxj6nm1d-kinit-5.31.0/bin:/nix/store/40b3czpgrs222fadr11y22phxxj6nm1d-kinit-5.31.0/lib/lib
    exec:/nix/store/40b3czpgrs222fadr11y22phxxj6nm1d-kinit-5.31.0/lib/libexec/kf5:/nix/store/mfran48hv83bmi9qiiklm0952zq45bf3-kserv
    ice-5.31.0/bin:/nix/store/cmypgamykrfys6aa1mz8zdzfqz3vcjv5-plasma-workspace-5.8.6/bin:/nix/store/cmypgamykrfys6aa1mz8zdzfqz3vcj
    v5-plasma-workspace-5.8.6/lib/libexec:/nix/store/1qc099pc7iydmj7ly2b1l73k3rs7vpn9-xmessage-1.0.4/bin:/nix/store/bkjji24z8prs7kp
    4ckddfzbz8h8j02m0-xprop-1.2.2/bin:/nix/store/n52jjwm629mbfk5k6lcp7p065139wqwv-xsetroot-1.1.0/bin:/nix/store/pxwylyb489aliqql7mz
    58xshh2cs8m52-yakuake-3.0.2/bin)

    ! %  ~/Playground  nix-shell ./fishie-shell.nix                                              Sat 25 Mar 2017 08:54:43 AM MST

    [nix-shell:~/Playground]$ killall
    Usage: killall [OPTION]... [--] NAME...
          killall -l, --list
          killall -V, --version

      -e,--exact          require exact match for very long names
      -I,--ignore-case    case insensitive process name match
      -g,--process-group  kill process group instead of process
      -y,--younger-than   kill processes younger than TIME
      -o,--older-than     kill processes older than TIME
      -i,--interactive    ask for confirmation before killing
      -l,--list           list all known signal names
      -q,--quiet          don't print complaints
      -r,--regexp         interpret NAME as an extended regular expression
      -s,--signal SIGNAL  send this signal instead of SIGTERM
      -u,--user USER      kill only process(es) running as USER
      -v,--verbose        report if the signal was successfully sent
      -V,--version        display version information
      -w,--wait           wait for processes to die

    [nix-shell:~/Playground]$ fish
    Linux davetim 4.9.17 x86_64
    08:54am  up   9:20,  1 user,  load average: 0.46, 0.35, 0.31
    ~/Playground  killall                                                                         Sat 25 Mar 2017 08:54:54 AM MST
    Usage: killall [OPTION]... [--] NAME...
          killall -l, --list
          killall -V, --version

      -e,--exact          require exact match for very long names
      -I,--ignore-case    case insensitive process name match
      -g,--process-group  kill process group instead of process
      -y,--younger-than   kill processes younger than TIME
      -o,--older-than     kill processes older than TIME
      -i,--interactive    ask for confirmation before killing
      -l,--list           list all known signal names
      -q,--quiet          don't print complaints
      -r,--regexp         interpret NAME as an extended regular expression
      -s,--signal SIGNAL  send this signal instead of SIGTERM
      -u,--user USER      kill only process(es) running as USER
      -v,--verbose        report if the signal was successfully sent
      -V,--version        display version information
      -w,--wait           wait for processes to die

But my nix-shell-fu is weak. What's the most natural way to have the shell drop to fish instead of bash as soon as you enter it? Is it some neat little shebang trick?

**Comment by xeji** [comment]
fixed in #45784
